### PR TITLE
Migrate to null-safety

### PR DESCRIPTION
**Pull request by nstrelow**
I previously started improving @RodrigoSMarques initial null-safety work, but I am very uncertain, if I can make all changes consistent throughout the code.

Therefore I am starting a new migration (mostly using nullable instead of empty String).  I am cross-checking all the suggestion from the dart migration tool with the decisions of @RodrigoSMarques.

I am adding annotations, where I disagree with the suggestions of the dart migration tool.

I will try to make the commits as small as possible, so the review can be easier.

I would love if somebody could have a look when it's finished.

List of already checked files:

### Dart SDK
- [x] lib/parse_server_sdk.dart 
- [x] lib/src/base/parse_constants.dart 
- [x] lib/src/data/parse_core_data.dart 
- [x] lib/src/data/parse_subclass_handler.dart 
- [x] lib/src/network/dio_adapter_io.dart 
- [x] lib/src/network/options.dart 
- [x] lib/src/network/parse_client.dart 
- [x] lib/src/network/parse_dio_client.dart 
- [x] lib/src/network/parse_http_client.dart 
- [x] lib/src/network/parse_query.dart
- [x] lib/src/network/parse_websocket_html.dart 
- [x] lib/src/network/parse_websocket_io.dart 
- [x] lib/src/network/parse_live_query.dart
- [x] lib/src/objects/parse_acl.dart
- [x] lib/src/objects/parse_base.dart  
- [x] lib/src/objects/parse_response.dart 
- [x] lib/src/objects/parse_object.dart 
- [x] lib/src/objects/parse_config.dart
- [x] lib/src/objects/parse_error.dart
- [x] lib/src/objects/parse_relation.dart
- [x] lib/src/objects/response/parse_error_response.dart
- [x] lib/src/objects/response/parse_response_utils.dart
- [x] lib/src/storage/core_store.dart 
- [x] lib/src/utils/parse_date_format.dart
- [x] test/parse_query_test.dart
- [x] lib/src/utils/parse_utils.dart
- [x] lib/src/objects/parse_function.dart
- [x] lib/src/objects/parse_session.dart
- [x] lib/src/utils/parse_logger.dart 
- [x] parse_file_extensions.dart: Deleted because deprecated
- [x] lib/src/utils/parse_decoder.dart
- [x] lib/src/utils/parse_encoder.dart
- [x] lib/src/objects/response/parse_response_builder.dart 
- [x] lib/src/storage/core_store_sem_impl.dart
- [x] lib/src/objects/parse_file_base.dart
- [x] lib/src/objects/parse_file.dart
- [x] lib/src/objects/parse_file_web.dart
- [x] lib/src/objects/parse_installation.dart
- [x] lib/src/objects/parse_user.dart 
- [x] lib/src/utils/parse_live_list.dart 

### Flutter SDK
- [x] lib/parse_server_sdk.dart 
- [x] lib/src/storage/core_store_sp_impl.dart
- [x] lib/src/utils/parse_live_grid.dart
- [x] lib/src/utils/parse_live_list.dart
- [x] test/parse_query_test.dart

### CODE REVIEW

**Comment by RodrigoSMarques** [comment]
@nstrelow 

Are you doing the review based on my Branch or the master branch of the repository?

**Comment by nstrelow** [comment]
I am not able to click on "Apply migration".

I have no clue what is wrong, it really sucks.

**Comment by RodrigoSMarques** [comment]
> I started from scratch based on the development branch of the official repo.
> 
> I am almost finished going through the whole dart sdk. Just LiveQueryList missing.
> 
> Then the flutter sdk next. Do you think I should put the flutter SDK in separate branch based on this one?

You can do it on the same branch. No problem.

**Comment by RodrigoSMarques** [comment]
@nstrelow 
I will ignore my changes and only work on yours.

You are doing an excellent job, doing a deeper analysis of the code and doing a general organization.

Thank you for taking the time to work on this and for support.

When you are done I will run my test project to validate.

**Comment by nstrelow** [comment]
@RodrigoSMarques Thank you very much.
Sadly I cannot run `Apply migration` with the changes I made. Dart migration tool just crashes.

If you could please try it out on your machine, that would be awesome. Just the `packages/dart`

**Comment by nstrelow** [comment]
Worked out the problem by using the commandline

`dart migrate --apply-migration`. Let's see if it works.

Flutter SDK (5 files) still needs migration. Will do that soon

**Comment by nstrelow** [comment]
The migration should be done.

The example apps have not been migrated yet.

**Comment by nstrelow** [comment]
@RodrigoSMarques @fischerscode 

I have not yet migrated the examples. @RodrigoSMarques you deleted them previously, should we delete them here as well?
Maybe you have a better example app?

**Comment by RodrigoSMarques** [comment]
> The migration should be done.
> 
> The example apps have not been migrated yet.

Thank you for the work done.

The sample apps, I don't see it as a priority now.

**Comment by RodrigoSMarques** [comment]
> @RodrigoSMarques @fischerscode
> 
> I have not yet migrated the examples. @RodrigoSMarques you deleted them previously, should we delete them here as well?
> Maybe you have a better example app?

I excluded the examples from my branch, as they had many errors during the analysis.

In the last commit they are back, but migrating them is not a priority.

https://github.com/RodrigoSMarques/Parse-SDK-Flutter/tree/nullsafety/packages/flutter


I'm going to merge the code and start testing with my validation application and then run on a real project that I have here.

@fischerscode 
Can you help with testing on any production app?

**Comment by nstrelow** [comment]
@RodrigoSMarques Awesome, thanks for merging!

**Comment by fischerscode** [comment]
@RodrigoSMarques 
> Can you help with testing on any production app?

I have an app, that just became production. I'll switch it to null safety soon.

File: packages/dart/lib/src/storage/core_store_sem_impl.dart [hunk 1]
```diff
@@ -30,20 +30,22 @@ class CoreStoreSembastImp implements CoreStore {
         return true;
       }());
       final Database db = await factory.openDatabase(dbPath,
-          codec: !parseIsWeb ? getXXTeaSembastCodec(password: password) : null);
+          codec: !parseIsWeb
+              ? getXXTeaSembastCodec(password: password = 'flutter_sdk')
```
Comment (RodrigoSMarques) [inline packages/dart/lib/src/storage/core_store_sem_impl.dart:34]: in a quick analysis, it looks like there is an error here.

even if the `password` is entered, the value will always be `flutter_sdk`

Comment (nstrelow) [reply packages/dart/lib/src/storage/core_store_sem_impl.dart:34]: Will check and report back

Comment (nstrelow) [reply packages/dart/lib/src/storage/core_store_sem_impl.dart:34]: It is a very weird the way it says `password = 'flutter_sdk'`

I would correct it to `password ?? 'flutter_sdk'`, which is probably what I meant to do.

Comment (nstrelow) [reply packages/dart/lib/src/storage/core_store_sem_impl.dart]: Good catch!

Ideally I would like to define it in the parameters, like this:
`func({String? password = 'flutter_sdk'})` (e.g. getInstance)
and I would like it to use the default value, when `func(password: null)` or when calling without arguments `func()`.

But dart does not work that way.

I will provide an example, as I found this quite confusing.

Comment (nstrelow) [reply packages/dart/lib/src/storage/core_store_sem_impl.dart]: Here is an example of how dart works (and what I do not like about it)

https://dartpad.dev/338245c6ada3f75690e2e24493387eb7?null_safety=true

Comment (nstrelow) [reply packages/dart/lib/src/storage/core_store_sem_impl.dart]: Another fix would be to mark the password as `required` and remove the default password.

But that might break some apps.

**Comment by pastordee** [comment]
Hi guys I'm trying to test the Nullsafety with my app but I keep getting this error,

Thanks



Note: Recompile with -Xlint:deprecation for details.
../../../development/flutter/.pub-cache/git/Parse-SDK-Flutter-123876a10d997f5bded788a0ee3388c0ecb736a3/packages/dart/lib/src/objects/response/parse_exception_response.dart:6:20: Error: 'DioError' isn't a type.
  if (exception is DioError) {
                   ^^^^^^^^
../../../development/flutter/.pub-cache/git/Parse-SDK-Flutter-123876a10d997f5bded788a0ee3388c0ecb736a3/packages/dart/lib/src/objects/response/parse_exception_response.dart:15:7: Error: Non-nullable variable 'errorResponse' must be assigned before it can be used.
      errorResponse ??= <String, dynamic>{};
      ^^^^^^^^^^^^^
../../../development/flutter/.pub-cache/git/Parse-SDK-Flutter-123876a10d997f5bded788a0ee3388c0ecb736a3/packages/dart/lib/src/objects/response/parse_exception_response.dart:15:7: Warning: Operand of null-aware operation '??=' has type 'Map<String, dynamic>' which excludes null.
 - 'Map' is from 'dart:core'.
      errorResponse ??= <String, dynamic>{};

      ^

**Comment by pastordee** [comment]
I'm also getting this (type 'Future<CoreStore>' is not a subtype of type 'FutureOr<CoreStoreSembastImp?>' in type cast
)

**Comment by RodrigoSMarques** [comment]
@pastordee 

Yes. It is not yet released for testing.

**Comment by pastordee** [comment]
> @pastordee
> 
> Yes. It is not yet released for testing.
> 
> Wait.

Oh okay thanks I thought we were meant to be testing it to see how it's working but I will wait.

**Comment by RodrigoSMarques** [comment]
@nstrelow 

Can you check?
I started analyzing this morning, but I couldn't finish it.
I am at work and will only have time at night.

This line is also in error.
Captura de Tela 2021-03-22 às 09 17 44

`dbPath` is not necessary for Flutter (only Dart).

**Comment by nstrelow** [comment]
@pastordee Yes, thanks for testing. I will see, if I can find the code which has those errors.

Do you have code snippets that produce that error?

@RodrigoSMarques Will look into it.

**Comment by pastordee** [comment]
> @pastordee Yes, thanks for testing. I will see, if I can find the code which has those errors.
> 
> Do you have code snippets that produce that error?
> 
> @RodrigoSMarques Will look into it.

ParseResponse buildParseResponseWithException(Exception exception) {
  final ParseResponse response = ParseResponse();
  if (exception is dio.DioError) {
    try {
      Map<String, dynamic> errorResponse;

      try {
        errorResponse =
            json.decode(exception.response?.data?.toString() ?? '{}');
      } catch (_) {}

      errorResponse ??= <String, dynamic>{};

      response.error = ParseError(
        message: errorResponse['error']?.toString() ??
            exception.response?.statusMessage,
        exception: exception,
        code: errorResponse['code'] ?? exception.response?.statusCode,
      );
    } catch (error) {
      response.error = ParseError(
          message: "Failed to build ParseResponse with exception",
          exception: error);
    }
  } else {
    response.error =
        ParseError(message: exception.toString(), exception: exception);
  }
  return response;
}

type 'Future' is not a subtype of type 'FutureOr<CoreStoreSembastImp?>' in type cast
s when I return the  CoreStoreSembastImp.getInstance(password: '+++++++++');

**Comment by nstrelow** [comment]
@pastordee I meant the code you are trying to run. Not the SDK code.

But I see the error, I just do not know, why it did not error for me. Very strange.

**Comment by nstrelow** [comment]
@RodrigoSMarques those errors were introduced by this: https://github.com/parse-community/Parse-SDK-Flutter/commit/de2c73d598d5b068ca1ffb42e78a60df3635b573

A weird merge merge commit. Was that intentional? What did the commit add? The naming is sadly only "Merge something"

**Comment by RodrigoSMarques** [comment]
> @pastordee I meant the code you are trying to run. Not the SDK code.
> 
> But I see the error, I just do not know, why it did not error for me. Very strange.

I also had the same problem when trying to run.

**Comment by RodrigoSMarques** [comment]
> @RodrigoSMarques those errors were introduced by this: [de2c73d](https://github.com/parse-community/Parse-SDK-Flutter/commit/de2c73d598d5b068ca1ffb42e78a60df3635b573)
> 
> A weird merge merge commit. Was that intentional? What did the commit add? The naming is sadly only "Merge something"

This is very strange.

In the tag of version 2.1.o which is the last version in production, the code looks like this:

[https://github.com/parse-community/Parse-SDK-Flutter/blob/2.1.0/packages/dart/lib/src/objects/response/parse_exception_response.dart](https://github.com/parse-community/Parse-SDK-Flutter/blob/2.1.0/packages/dart/lib/src/objects/response/parse_exception_response.dart)
In the master branch, which should be with the same version, it is different. 
[https://github.com/parse-community/Parse-SDK-Flutter/blob/master/packages/dart/lib/src/objects/response/parse_exception_response.dart](https://github.com/parse-community/Parse-SDK-Flutter/blob/master/packages/dart/lib/src/objects/response/parse_exception_response.dart)

There was a problem with the merge.

**Comment by RodrigoSMarques** [comment]
> Oh okay thanks I thought we were meant to be testing it to see how it's working but I will wait.

Everything is fine.
I’ll run some tests and let you know when it’s ready.

**Comment by nstrelow** [comment]
I branched off of development not master.

**Comment by nstrelow** [comment]
I am improving said code with DioError.

**Comment by nstrelow** [comment]
https://github.com/parse-community/Parse-SDK-Flutter/pull/589

Should now be fixed. Please be careful with merges next time @RodrigoSMarques 😁

**Comment by RodrigoSMarques** [comment]
> I branched off of development not master.

Everyone should be the same:

- [Tag 2.1](https://github.com/parse-community/Parse-SDK-Flutter/blob/2.1.0/packages/dart/lib/src/objects/response/parse_exception_response.dart):  
- [Branch: Development](https://github.com/parse-community/Parse-SDK-Flutter/blob/development/packages/dart/lib/src/objects/response/parse_exception_response.dart)
- [Branch: Master](https://github.com/parse-community/Parse-SDK-Flutter/blob/master/packages/dart/lib/src/objects/response/parse_exception_response.dart)

**Comment by RodrigoSMarques** [comment]
> Should now be fixed. Please be careful with merges next time @RodrigoSMarques 😁

#566 PR from tag 2.1.0 to Master
This PR contains the #482  fix to DioError

Nullsafety branch was created from the master branch

**Comment by RodrigoSMarques** [comment]
@nstrelow 

What version of Flutter are you running? Stable channel?

I can't run any mobile project with your PR.

`FutureOr` is not accepted. Error during execution.

I had to undo changes, in the CoreStoreSembastImp class of the Dart package

Changes:

Captura de Tela 2021-03-23 às 01 25 46

Captura de Tela 2021-03-23 às 01 28 11


Cloud Function call failed:
```dart
Failed Call Cloud Function: Unsupported operation: Cannot set value in unmodifiable Map
```

**Comment by RodrigoSMarques** [comment]
> @nstrelow
> 
> What version of Flutter are you running? Stable channel?
> 
> I can't run any mobile project with your PR.
> 
> `FutureOr` is not accepted. Error during execution.
> 
> I had to undo changes, in the CoreStoreSembastImp class of the Dart package
> 
> Changes:
> 
> Captura de Tela 2021-03-23 às 01 25 46
> 
> Captura de Tela 2021-03-23 às 01 28 11
> 
> Cloud Function call failed:
> 
> ```dart
> Failed Call Cloud Function: Unsupported operation: Cannot set value in unmodifiable Map
> ```

Cloud Function Fix

**Comment by nstrelow** [comment]
Thanks for testing, will look into it 😁

**Comment by RodrigoSMarques** [comment]
> Thanks for testing, will look into it 😁
> 
> I used the shared preference store and probably was to tired to think over other things to test. Sorry for that.

No problem. I will submit the PR

**Pull request merged**